src/Toml/Parser/TOML.hs

```diff
@@ -1,44 +1,46 @@
 module Toml.Parser.TOML
-       ( keyValP
+       ( hasKeyP
        , tableHeaderP
+       , inlineTableP
        , tomlP
        ) where
 
 import Control.Applicative (Alternative (many))
+import Control.Applicative.Combinators (sepEndBy, between, eitherP)
 
 import Toml.Parser.Core (Parser, sc, text)
-import Toml.Parser.Value (keyP, tableNameP, valueP)
+import Toml.Parser.Value (keyP, tableNameP, anyValueP)
 import Toml.PrefixTree (Key (..), fromList)
-import Toml.Type (AnyValue, TOML (..), typeCheck)
+import Toml.Type (AnyValue, TOML (..))
 
 import qualified Data.HashMap.Lazy as HashMap
 
+hasKeyP :: Parser (Key, Either AnyValue TOML)
+hasKeyP = (,) <$> keyP <* text "=" <*> eitherP anyValueP inlineTableP
 
-keyValP :: Parser (Key, AnyValue)
-keyValP = do
-    k    <- keyP
-    _    <- text "="
-    uval <- valueP
-    case typeCheck uval of
-        Left err -> fail $ show err
-        Right v  -> pure (k, v)
-
+inlineTableP :: Parser TOML
+inlineTableP = between (text "{") (text "}")
+                       (makeToml <$> hasKeyP `sepEndBy` text ",")
 
 tableHeaderP :: Parser (Key, TOML)
-tableHeaderP = do
-    k    <- tableNameP
-    toml <- makeToml <$> many keyValP
-    pure (k, toml)
+tableHeaderP = (,) <$> tableNameP <*> (makeToml <$> many hasKeyP)
+
+distributeEithers :: [(c, Either a b)] -> ([(c, a)], [(c, b)])
+distributeEithers = foldr distribute ([], [])
```

Comment:
I wonder whether usage of `foldr` is okay here and whether it leaks memory or not 🤔

Reply:
My first instinct was to use `Data.List.partition :: (a -> Bool) -> [a] -> ([a],[a])` but I wanted to use pattern matching. Also, looking at it now, it's implemented with `foldr`. 
I don't know anything about memory leaks, but I'm pretty sure `foldr` is used a lot. Also, I doubt there would be toml files with that many keys to cause trouble.

Reply:
@jiegillet Yeah, `foldr` looks okay in this case, it's just not obvious whether it's totally okay. `foldr` can introduce space leaks, so it should be used with care. I also don't expect the tables to be big, so it's okay to leave it as it is now.